sklearn/cluster/_kmeans.py

```diff
@@ -1654,121 +1631,94 @@ def fit(self, X, y=None, sample_weight=None):
-
-            # Compute the label assignment on the init dataset
-            _mini_batch_step(
-                X_valid, sample_weight_valid,
-                x_squared_norms[validation_indices], cluster_centers,
-                weight_sums, old_center_buffer, False, distances=None,
-                verbose=self.verbose)
-
-            # Keep only the best cluster centers across independent inits on
-            # the common validation set
-            _, inertia = _labels_inertia(X_valid, sample_weight_valid,
-                                         x_squared_norms_valid,
-                                         cluster_centers)
+                X, x_squared_norms=x_squared_norms, init=init,
+                random_state=random_state, init_size=self._init_size)
+
+            # Compute inertia on a validation set.
+            _, inertia = _labels_inertia(
+                X_valid, sample_weight_valid, x_squared_norms_valid,
+                cluster_centers, n_threads=self._n_threads)
+
             if self.verbose:
-                print("Inertia for init %d/%d: %f"
-                      % (init_idx + 1, self._n_init, inertia))
+                print(f"Inertia for init {init_idx + 1}/{self._n_init}: "
+                      f"{inertia}")
             if best_inertia is None or inertia < best_inertia:
-                self.cluster_centers_ = cluster_centers
-                self._counts = weight_sums
+                init_centers = cluster_centers
                 best_inertia = inertia
 
-        # Empty context to be used inplace by the convergence check routine
-        convergence_context = {}
+        centers = init_centers
+        centers_new = np.empty_like(centers)
+
+        # Initialize counts
+        self._counts = np.zeros(self.n_clusters, dtype=X.dtype)
+
+        # Attributes to monitor the convergence
+        self._ewa_inertia = None
+        self._ewa_inertia_min = None
+        self._no_improvement = 0
 
-        # Perform the iterative optimization until the final convergence
-        # criterion
-        for iteration_idx in range(n_iter):
+        n_batches = int(np.ceil(float(n_samples) / self.batch_size))
+        n_iter = int(self.max_iter * n_batches)
+
+        # Perform the iterative optimization until convergence
+        for i in range(n_iter):
             # Sample a minibatch from the full dataset
-            minibatch_indices = random_state.randint(
-                0, n_samples, self.batch_size)
+            minibatch_indices = random_state.randint(0, n_samples,
+                                                     self.batch_size)
+
+            # Randomly choose whether to perform random reassignment:
+            # the choice is done as a function of the iteration index, and the
+            # minimum number of counts, in order to force this reassignment to
+            # happen every once in a while.
+            random_reassign = (i + 1) % (10 + int(self._counts.min())) == 0
```

Comment:
I cannot remember the motivation for this expression and it's not intuitive why it does it this way by reading it.

Actually, in `partial_fit` it's implemented as:

```python
            # The lower the minimum count is, the more we do random
            # reassignment, however, we don't want to do random
            # reassignment too often, to allow for building up counts
            random_reassign = self._random_state.randint(
                10 * (1 + self._counts.min())) == 0
```

which makes more sense.

But I suspect we could factorize this logic into a private method to be reused in both methods, and maybe make it evolve to a simple heuristic that takes into account explicitly the number of training example seen at the time of the last effective reassignment event.